pymeasure/instruments/lecroy/lecroyT3DSO1204.py

```diff
@@ -756,40 +796,123 @@ def download_image(self):
+        # to get the number of sampled points
+        if self.waveform_source == "MATH":
+            sample_points = None
+        else:
+            sample_points = getattr(self, f"acquisition_sample_size_{self.waveform_source.lower()}")
+
+        # Check that we are trying to read a reasonable amount of points
+        if num_points is not None and sample_points is not None and num_points > sample_points:
+            ValueError(f"Number of requested points ({num_points}) greater than "
+                       f"number of sampled points ({sample_points})")
+
+        # Setup waveform acquisition parameters
         source = _sanitize_source(source)
         self.waveform_source = source
-        if points is not None:
-            self.waveform_points = points
+        self.waveform_first_point = 0
+        self.waveform_points = 0
+        self.waveform_sparsing = 0
+        if num_points is not None:
+            self.waveform_points = num_points
         if sparsing is not None:
             self.waveform_sparsing = sparsing
-        if first_point is not None:
-            self.waveform_first_point = first_point
 
+        # Read waveform preamble (waveform metadata)
         preamble = self.waveform_preamble
-        data_bytes = self._digitize(source)
-        return np.array(data_bytes), preamble
+
+        # Check how many points are to be expected
+        self.write(f"{source}:WF? DAT2")
+        values = self._digitize(source=source)
+        expected_points = _header_sanity_checks(values)
+
+        # If the number of points is big enough, split the data in small chunks and read it one
+        # chunk at a time. For less than 100K points we do not bother splitting them.
+        chunk_bytes = 1e5 if expected_points < 1e5 else 1e6
+        chunk_points = chunk_bytes - self._header_size - self._footer_size
+        iterations = _ceildiv(expected_points, chunk_points)
+        i = 0
+        data = []
+        while True:
+            first_point = i * chunk_points
+            remaining_points = expected_points - first_point
+            if remaining_points > chunk_points:
+                requested_points = chunk_points
+            else:
+                requested_points = remaining_points
+            requested_bytes = requested_points + self._header_size + self._footer_size
+            self.waveform_first_point = first_point
+            self.waveform_points = requested_points
+            values = self._digitize(source=source, requested_bytes=requested_bytes)
+            _header_sanity_checks(values)
+            _footer_sanity_checks(values)
+            _npoints_sanity_checks(values)
+            data.append(values[16:-2])
+            i += 1
+            if i >= iterations:
+                break
+        data = np.concatenate(data)
+
+        def _scale_data(x):
+            value = bitstring.Bits(uint=int(x), length=8).unpack('int')[0] * preamble["ydiv"] / 25.
```

Comment:
We have to talk about adding a new dependency. What are you trying to do here?

Reply:
I am trying to convert from a raw byte to a signed int. Of course it can be done without the bitstring module. 

For further details refer to the top of page 181 of the oscilloscope [programming guide](https://evision-webshop.de/WebRoot/Store13/Shops/e6077b56-ef34-48be-850f-07696838e42e/MediaGallery/Siglent/Siglent-Programming-Guide-2.pdf). However there is a error in the text and the example should be `252-256 = -4` instead of `252-255 = -3`.

This is the same that Bastian is doing [here](https://github.com/bleykauf/pymeasure/blob/9faeac6a20d48fb2386123350115ebd64e9fceef/pymeasure/instruments/teledyne/t3ds01204.py#L277).


PS Either way is fine with me but the bitstring module is very widely used and readily available so it should not be too much trouble to add.

Reply:
`int.from_bytes` is able to return an integer from a bytes object. I have to look more into the code to see, whether it works here or not.

Reply:
`int.from_bytes([x], byteorder="big", signed=True)` turns the byte with value `x` into a signed integer. Alternatively, you could use bitmasks: `value -= (value & 128) << 1`.

Reply:
This is using ` int.from_bytes` now.

Reply:
Yes, at some point I applied Benedikt's suggestion.